src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -192,458 +96,48 @@ void IUtilities::updateImageViaStagingBuffer(
-                if(currentBlockInRow == 0 && currentRowInSlice == 0 && currentSliceInLayer == 0 && remainingLayersInRegion > 0)
-                {
-                    bool success = tryFillRegion();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
-                }
-                // There is remaining slices in layer that needs copying
-                else if (currentBlockInRow == 0 && currentRowInSlice == 0 && currentSliceInLayer > 0)
-                {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillLayer();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
-                }
-                // There is remaining rows in slice that needs copying
-                else if (currentBlockInRow == 0 && currentRowInSlice > 0)
-                {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillSlice();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
-                }
-                // There is remaining blocks in row that needs copying
-                else if (currentBlockInRow > 0)
+                asset::IImage::SBufferCopy nextRegionToCopy = {};
+                if(availableUploadBufferMemory > 0u && regionIterator.advanceAndCopyToStagingBuffer(nextRegionToCopy, availableUploadBufferMemory, currentUploadBufferOffset, m_defaultUploadBuffer->getBufferPointer()))
                 {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillRow();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
+                    anyTransferRecorded = true;
+                    regionsToCopy.push_back(nextRegionToCopy);
                 }
+                else
+                    break;
             }
 
             if(!regionsToCopy.empty())
             {
                 cmdbuf->copyBufferToImage(m_defaultUploadBuffer.get()->getBuffer(), dstImage, dstImageLayout, regionsToCopy.size(), regionsToCopy.data());
             }
 
-            assert(anyTransferRecorded && "uploadBufferSize is not enough to support the smallest possible transferable units to image, may be caused if your queueFam's minImageTransferGranularity is large or equal to <0,0,0>.");
+            assert(anyTransferRecorded && "allocationSize is not enough to support the smallest possible transferable units to image, may be caused if your queueFam's minImageTransferGranularity is large or equal to <0,0,0>.");
```

Comment:
use the logger, also it is possible to validate up-front the queue granularity stuff.

> Possible values of minImageTransferGranularity are:

(0,0,0) specifies that only whole mip levels must be transferred using the image transfer operations on the corresponding queues. In this case, the following restrictions apply to all offset and extent parameters of image transfer operations:

The x, y, and z members of a [VkOffset3D](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkOffset3D.html) parameter must always be zero.

The width, height, and depth members of a [VkExtent3D](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkExtent3D.html) parameter must always match the width, height, and depth of the image subresource corresponding to the parameter, respectively.

(Ax, Ay, Az) where Ax, Ay, and Az are all integer powers of two. In this case the following restrictions apply to all image transfer operations:

x, y, and z of a [VkOffset3D](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkOffset3D.html) parameter must be integer multiples of Ax, Ay, and Az, respectively.

width of a [VkExtent3D](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkExtent3D.html) parameter must be an integer multiple of Ax, or else x + width must equal the width of the image subresource corresponding to the parameter.

height of a [VkExtent3D](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkExtent3D.html) parameter must be an integer multiple of Ay, or else y + height must equal the height of the image subresource corresponding to the parameter.

depth of a [VkExtent3D](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkExtent3D.html) parameter must be an integer multiple of Az, or else z + depth must equal the depth of the image subresource corresponding to the parameter.

If the format of the image corresponding to the parameters is one of the block-compressed formats then for the purposes of the above calculations the granularity must be scaled up by the compressed texel block dimensions.

Reply:
Ok
I already validate all these but in the wrong place due to copy pasting shit around.
I'll need to take logger as IUtilities input,

Reply:
> Ok I already validate all these but in the wrong place due to copy pasting shit around.

Most of this is already validated (just not granularity) in `IImage::validateCopies`, so maybe extend that.

> I'll need to take logger as IUtilities input,

it either already has it somewhere, or you can make it take one in the ctor.